modules/react/text-input/lib/InputGroup.tsx

```diff
@@ -0,0 +1,155 @@
+import React from 'react';
+
+import {space} from '@workday/canvas-kit-react/tokens';
+import {createComponent, ExtractProps, useIsRTL} from '@workday/canvas-kit-react/common';
+import {Flex} from '@workday/canvas-kit-react/layout';
+
+import {TextInput} from './TextInput';
+
+export const InputGroupStart = createComponent('div')({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return (
+      <Flex
+        ref={ref}
+        as={Element}
+        position="absolute"
+        alignItems="center"
+        justifyContent="center"
+        height="xl"
+        width="xl"
+        {...elemProps}
+      />
+    );
+  },
+});
+
+export const InputGroupEnd = createComponent('div')({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return (
+      <Flex
+        ref={ref}
+        as={Element}
+        position="absolute"
+        alignItems="center"
+        justifyContent="center"
+        height="xl"
+        width="xl"
+        {...elemProps}
+      />
+    );
+  },
+});
+
+export const InputGroupInput = createComponent(TextInput)({
+  Component(elemProps: ExtractProps<typeof Flex>, ref, Element) {
+    return <Flex ref={ref} as={Element} width="100%" {...elemProps} />;
+  },
+});
+
+// make sure we always use pixels if the input is a number - this is required for `calc`
+const toPx = (input: string | number): string => {
+  return typeof input === 'number' ? `${input}px` : input;
+};
+
+/**
+ * An `InputGroup` is a container around a {@link TextInput} with optional start and end elements.
+ * The start and end elements are usually icons or icon buttons. The `InputGroup` will add padding
+ * to the input so the icons/buttons display correctly. This component uses `React.Children.map` and
+ * `React.cloneElement` from the [React.Children](https://react.dev/reference/react/Children) API.
+ * This means all children must be `InputGroup.*` components. Any other direct children will cause
+ * issues. You can add different elements/components inside the
+ * {@link InputGroupStart InputGroup.Start} and {@link InputGroupEnd InputGroup.End} subcomponents.
+ *
+ * ```tsx
+ * <InputGroup>
+ *   <InputGroup.Start as={SystemIcon} pointerEvents="none" icon={searchIcon} />
+ *   <InputGroup.Input />
+ *   <InputGroup.End>
+ *     <TertiaryButton tabIndex={-1} icon={xIcon} size="small" />
+ *   </InputGroup.End>
+ * </InputGroup>
+ * ```
+ */
+export const InputGroup = createComponent('div')({
+  displayName: 'InputGroup',
+  Component({children, ...elemProps}: ExtractProps<typeof Flex>, ref, Element) {
+    const isRTL = useIsRTL();
+    let paddingInlineStart: string;
+    let paddingInlineEnd: string;
+    const offsetsStart: string[] = [];
+    const offsetsEnd: string[] = [];
+
+    React.Children.forEach(children, child => {
```

Comment:
This is offsetting the padding based on if there's an element at the start or end so that the input has the correct padding/size instead of it being overlapped by the position absolute elements.

Reply:
I added a comment

Reply:
WWMA